File: app/src/main/java/hm/orz/chaos114/android/slideviewer/widget/RowSelectOcrLanguageView.java
package hm.orz.chaos114.android.slideviewer.widget;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.SwitchCompat;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.LinearLayout;

import hm.orz.chaos114.android.slideviewer.R;
import hm.orz.chaos114.android.slideviewer.databinding.RowSelectOcrLanguageBinding;
import hm.orz.chaos114.android.slideviewer.infra.repository.SettingsRepository;
import hm.orz.chaos114.android.slideviewer.ocr.model.Language;
import hm.orz.chaos114.android.slideviewer.ocr.util.DirectorySettings;
import lombok.Setter;

/**
 * List row of select ocr language.
 */
public class RowSelectOcrLanguageView extends LinearLayout {

    private RowSelectOcrLanguageBinding binding;
    private Language language;
    private boolean loading;
    @Setter
    private RowSelectOcrLanguageViewListener listener;

    public RowSelectOcrLanguageView(Context context) {
        this(context, null);
    }

    public RowSelectOcrLanguageView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RowSelectOcrLanguageView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        binding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.row_select_ocr_language, this, true);
    }

    public void setData(Language lang) {
        this.language = lang;

        binding.label.setText(language.getLabel());
        if (loading) {
            binding.status.setText(R.string.select_ocr_language_downloading);
        } else {
            binding.status.setText(DirectorySettings.hasFile(getContext(), language)
                    ? R.string.select_ocr_language_downloaded
                    : R.string.select_ocr_language_not_downloaded);
        }
        SettingsRepository settingsRepository = new SettingsRepository(getContext());
        binding.languageSwitch.setChecked(language.getId().equals(settingsRepository.getSelectedLanguage()));

        binding.languageSwitch.setOnClickListener(v -> {
            if (listener != null) {
                listener.onChangeState(this, language, ((SwitchCompat) v).isChecked());
            }
        });
        binding.languageSwitch.setVisibility(loading ? GONE : VISIBLE);
        binding.progress.setVisibility(loading ? VISIBLE : GONE);
    }

    public void showLoading(boolean loading) {
        this.loading = loading;

        // update view
        setData(language);
    }

    public interface RowSelectOcrLanguageViewListener {
        void onChangeState(RowSelectOcrLanguageView view, Language language, boolean isChecked);
    }
}

File: app/src/main/java/hm/orz/chaos114/android/slideviewer/widget/RowSelectOcrLanguageView.kt
package hm.orz.chaos114.android.slideviewer.widget

import android.content.Context
import android.databinding.DataBindingUtil
import android.support.v7.widget.SwitchCompat
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout

import hm.orz.chaos114.android.slideviewer.R
import hm.orz.chaos114.android.slideviewer.databinding.RowSelectOcrLanguageBinding
import hm.orz.chaos114.android.slideviewer.infra.repository.SettingsRepository
import hm.orz.chaos114.android.slideviewer.ocr.model.Language
import hm.orz.chaos114.android.slideviewer.ocr.util.DirectorySettings

/**
 * List row of select ocr language.
 */
class RowSelectOcrLanguageView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    private val binding: RowSelectOcrLanguageBinding
            = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.row_select_ocr_language, this, true)

    private var language: Language? = null
    private var loading: Boolean = false
    private var listener: RowSelectOcrLanguageViewListener? = null

    fun setData(lang: Language) {
        this.language = lang

        binding.label.text = lang.label
        binding.status.setText(when {
            loading ->
                R.string.select_ocr_language_downloading
            DirectorySettings.hasFile(context, lang) ->
                R.string.select_ocr_language_downloaded
            else ->
                R.string.select_ocr_language_not_downloaded
        })
        // TODO inject by Dagger?
        val settingsRepository = SettingsRepository(context)
        binding.languageSwitch.isChecked = lang.id == settingsRepository.selectedLanguage

        binding.languageSwitch.setOnClickListener { v ->
            listener?.onChangeState(this, lang, (v as SwitchCompat).isChecked)
        }
        binding.languageSwitch.visibility = if (loading) View.GONE else View.VISIBLE
        binding.progress.visibility = if (loading) View.VISIBLE else View.GONE
    }

    fun showLoading(loading: Boolean) {
        this.loading = loading

        // update view
        setData(language!!)
    }

    fun setListener(listener: RowSelectOcrLanguageViewListener) {
        this.listener = listener
    }

    interface RowSelectOcrLanguageViewListener {
        fun onChangeState(view: RowSelectOcrLanguageView, language: Language, isChecked: Boolean)
    }
}
